File: 23.RadixSort.java
/* ----------------------------------------------------------------------------  */
/*   ( The Authentic JS/JAVA CodeBuff )
 ___ _                      _              _ 
 | _ ) |_  __ _ _ _ __ _ __| |_ __ ____ _ (_)
 | _ \ ' \/ _` | '_/ _` / _` \ V  V / _` || |
 |___/_||_\__,_|_| \__,_\__,_|\_/\_/\__,_|/ |
                                        |__/ 
 */
/* --------------------------------------------------------------------------   */
/*    Youtube: https://youtube.com/@code-with-Bharadwaj                        */
/*    Github : https://github.com/Manu577228                                  */
/*    Portfolio : https://manu-bharadwaj-portfolio.vercel.app/portfolio        */
/* -----------------------------------------------------------------------  */

import java.io.*;

public class RadixSortDemo {

    // Counting Sort for a given digit (exp = 1, 10, 100,...)
    static void countingSort(int[] arr, int exp) {
        int n = arr.length;
        int[] output = new int[n];     // Step 1: Temporary output array
        int[] count = new int[10];     // Step 2: Count array for digits 0-9

        // Step 3: Count occurrences of each digit
        for (int i = 0; i < n; i++) {
            int index = (arr[i] / exp) % 10;
            count[index]++;
        }

        // Step 4: Convert count to cumulative count
        for (int i = 1; i < 10; i++) {
            count[i] += count[i - 1];
        }

        // Step 5: Build the output array (stable - from end)
        for (int i = n - 1; i >= 0; i--) {
            int index = (arr[i] / exp) % 10;
            output[count[index] - 1] = arr[i];
            count[index]--;
        }

        // Step 6: Copy sorted output back to original array
        for (int i = 0; i < n; i++) {
            arr[i] = output[i];
        }
    }

    // Main Radix Sort
    static void radixSort(int[] arr) {
        // Step 7: Find max to know number of digits
        int max = arr[0];
        for (int i = 1; i < arr.length; i++) {
            if (arr[i] > max) max = arr[i];
        }

        // Step 8: Do counting sort for every digit
        for (int exp = 1; max / exp > 0; exp *= 10) {
            countingSort(arr, exp);
        }
    }

    public static void main(String[] args) throws Exception {
        int[] nums = {170, 45, 75, 90, 802, 24, 2, 66};

        System.out.print("Original: ");
        for (int x : nums) System.out.print(x + " ");
        System.out.println();

        radixSort(nums);

        System.out.print("Sorted: ");
        for (int x : nums) System.out.print(x + " ");
        System.out.println();
    }
}

File: 23.RadixSort.py
# Radix Sort is a non-comparative integer sorting algorithm that sorts data with 
# integer keys by grouping keys by individual digits (from least significant to most significant).
# It uses a stable sub-sorting algorithm like Counting Sort for each digit place.

# Explanation

# Radix Sort works digit by digit, starting from the least significant digit (LSD) to the most significant digit (MSD).

# For each digit position, it uses a stable sort (like counting sort) to order elements according to that digit.

# After processing all digits, the array becomes completely sorted.

# Time Complexity: O(d × (n + k)), where

# d = number of digits in the largest number

# n = number of elements

# k = range of digits (usually 0–9 for decimal numbers)

def counting_sort(arr, exp):
    n = len(arr)
    output = [0] * n
    count = [0] * 10

    for i in range(n):
        index = (arr[i] // exp) % 10
        count[index] += 1;

    for i in range(1, 10):
        count[i] += count[i - 1]

    for i in range(n - 1, -1, -1):
        index = (arr[i] // exp) % 10
        output[count[index] - 1] = arr[i]
        count[index] -= 1

    for i in range(n):
        arr[i] = output[i]

def radix_sort(arr):
    max_num = max(arr)

    exp = 1

    while max_num // exp > 0:
        counting_sort(arr, exp)
        exp *= 10

nums = [170, 45, 75, 90, 802, 24, 2, 66]

print("Original:", nums)
radix_sort(nums)
print("Sorted:", nums)

